Warn users if case default panels are outdated

northwestwitch opened the pull request:
Display a warning badge if case default panels' version is older than panel version in database. This PR goes together with #2007

![image](https://user-images.githubusercontent.com/28093618/92280947-db7f2900-eefa-11ea-9688-91512d9813f4.png)

**How to test**:
1. Can be done both on stage or a local implementation of scout.
1. Update a gene panel that is the default gene panel of a case
1. Go to the case and see the warning badge
1. On mouseover it should tell you the gene differences between case panel and updated panel.

**Expected outcome**:
The functionality should be working
Take a screenshot and attach or copy/paste the output.

**Review:**
- [ ] code approved by
- [ ] tests executed by

dnil commented:
Tremendously good! Now have a very good and deserved weekend!

northwestwitch commented:
You too! 😊

dnil commented:
Now this is certainly an edge case, but still:
![Screenshot 2020-09-07 at 11 20 35](https://user-images.githubusercontent.com/758570/92371738-4a77a000-f0fc-11ea-8bb5-18b89f79ada3.png)
Version 2 w/o POT1, version 3 with it again..

dnil commented on scout/server/blueprints/cases/controllers.py:
```diff
@@ -96,23 +96,45 @@ def case(store, institute_obj, case_obj):
+            extra_genes = [
+                gene.get("symbol", gene["hgnc_id"])
+                for gene in panel_obj["genes"]
+                if gene not in latest_panel["genes"]
```
So I guess the issue is here, right? The "genes" are fairly complex objects, and can easily differ over time? Should be something like comparing the actual gene["hgnc_id"] for each of those?

northwestwitch replied:
Do you want to compare only the HGNC id or the symbol and the HGNC ID?

dnil replied:
ID only: that is not supposed to change, whereas the gene symbol is quite likely to have changed for a few from the first version!

northwestwitch commented:
![image](https://user-images.githubusercontent.com/28093618/92375439-616cc100-f101-11ea-9d10-92f4f5ded147.png)


> Version 2 w/o POT1, version 3 with it again..

Weird, it's like if it doesn't update, nice find!

northwestwitch commented:
@dnil should be fixed now. It started complaining for the code complexity, so I moved the gene check to a separate function. It didn't work. I think it keeps complaining because the case  controller is still very long but it was like this from before..

dnil commented:
Function tests ok! Really nice one!

How about adding a text "The case panel is used in variants filter" to the tooltip. This can then be changed for #2007, to also state that "The latest panel version is used in variants filter".

dnil commented on scout/server/blueprints/cases/controllers.py:
```diff
@@ -96,23 +96,37 @@ def case(store, institute_obj, case_obj):
+
+        # Check if case-specific panel is up-to-date with latest version of the panel
+        if panel_obj["version"] < latest_panel["version"]:
+            extra_genes, missing_genes = _check_outdated_gene_panel(panel_obj, latest_panel)
```
:+1:

dnil commented on scout/server/blueprints/cases/controllers.py:
```diff
@@ -198,6 +212,39 @@ def case(store, institute_obj, case_obj):
+    ]
+    # Extract the genes unique to case panel
+    extra_genes = [
+        gene["symbol"] for gene in hgnc_ids_case_panel if gene not in hgnc_ids_latest_panel
```
Sets might have been even faster, but this doesn't seem bad either. Nicely done!

dnil replied:
> @dnil should be fixed now. It started complaining for the code complexity, so I moved the gene check to a separate function. It didn't work. I think it keeps complaining because the case controller is still very long but it was like this from before..

Well, yes, I agree it is complex, and I suppose what pushed it over the edge is the new nested if statement. Its possible you would get away with it if you just un-nest them - the vars will not be more or less set if you check them outside the outer if-statement I suppose.

dnil replied:
That is to say, I don't mind much either way, being c-schooled, but I suppose flat if-statements is more the python way.

northwestwitch replied:
> Sets might have been even faster, but this doesn't seem bad either. Nicely done!

The problem with sets here is that I'm comparing the hgnc ids, but I need to keep also the symbols, and to do so I store the values in dictionaries. Dictionaries can't be stored in sets (not hashable) so I'd need to have two loops instead, and in the end it would be slower I think. I'll keep it like this instead

northwestwitch replied:
> That is to say, I don't mind much either way, being c-schooled, but I suppose flat if-statements is more the python way.

I wouldn't like to add even more code to a function that by itself is simple. It's just that it's part of a nested check and the complexity is also given by the levels above. Let's keep it this way. We can fix it in a second moment if we feel like it.

northwestwitch commented:
> Function tests ok! Really nice one!
> 
> How about adding a text "The case panel is used in variants filter" to the tooltip. This can then be changed for #2007, to also state that "The latest panel version is used in variants filter".

I think it would be good to have a tooltip, but I'd rather add it to the variants search select instead. Otherwise this tooltip will become super long in the case there are many gene differences between the case panel and the latest panel

northwestwitch commented:
I think I'm going to merge this, and I'll add the tooltip that the filtering is performed using the latest panel on the variants filtering page, do you agree @dnil?

northwestwitch commented:
I changed my mind and added "Case panel version is used in variants filtering" to the case tooltip instead. Have you noticed how voluble I am in my decisions sometimes? 😆

dnil commented:
I'd call it a great flexibility of mind! One should not needlessly get stuck on one of many almost equally good solutions.. 😉
